llvm/lib/Transforms/Scalar/MemCpyOptimizer.cpp

```diff
@@ -399,6 +399,9 @@ Instruction *MemCpyOptPass::tryMergingIntoMemset(Instruction *StartInst,
       // If this is a store, see if we can merge it in.
       if (!NextStore->isSimple()) break;
 
+      if (AA->isNoAlias(NextStore->getPointerOperand(), StartPtr))
+        continue;
```

Comment:
Would be better to integrate this with the checks above by checking getModRef == NoModRef instead. That covers all instruction types.

Also, please use BatchAAResults here.